rust/connlib/tunnel/src/client.rs

```diff
@@ -1538,4 +1538,29 @@ mod proptests {
             expected_routes(vec![cidr_resource2.address])
         );
     }
+
+    #[test_strategy::proptest]
+    fn resources_can_be_replaced_with_same_id(
+        #[strategy(cidr_resource())] cidr_resource: ResourceDescriptionCidr,
+        #[strategy(ip_network())] new_ip: IpNetwork,
+        #[strategy(resource_name())] new_name: String,
+    ) {
+        let mut client_state = ClientState::for_test();
+        client_state.add_resources(&[ResourceDescription::Cidr(cidr_resource.clone())]);
+
+        let mut cidr_resource = cidr_resource.clone();
+        cidr_resource.address = new_ip;
+        cidr_resource.name = new_name;
+
+        client_state.set_resources(vec![ResourceDescription::Cidr(cidr_resource.clone())]);
+
+        assert_eq!(
+            hashset(client_state.resources().iter()),
+            hashset(&[ResourceDescription::Cidr(cidr_resource.clone()),])
+        );
+        assert_eq!(
+            hashset(client_state.routes()),
+            expected_routes(vec![cidr_resource.address])
+        );
```

Comment:
This is the exact same assertion as we already have above:

- We call `set_resources` with a resource
- We assert that we have this exact resource
- We assert that we have a route with this resource's address

This functionality is already covered by the previous test. In a way, this test is actually testing _less_ because it would not catch a bug that doesn't remove resources not present in the new set.

What you want to test here is that it doesn't get removed and added. But for this, we need to assert on state that gets destroyed as part of `remove_resource` (like connections to gateways if it was the last one).

Reply:
It's not the exact same setup though, it could be the case, that if the id of the resource in `set_resource` is the same as in the previous one, the address and name aren't changed.

The previous test would pass if we change the `set_resource` to:

```rs
    fn set_resources(&mut self, new_resources: Vec<ResourceDescription>) {
        self.remove_resources(
            &HashSet::from_iter(self.resource_ids.keys().copied())
                .difference(&HashSet::<ResourceId>::from_iter(
                    new_resources.iter().map(|r| r.id()),
                ))
                .copied()
                .collect_vec(),
        );

        self.add_resources(
            &new_resources
                .iter()
                .filter(|r| self.resource_ids.contains_key(&r.id()))
                .cloned()
                .collect_vec(),
        );
    }

```

but this one would fail

Reply:
Oh, actually testing with that it does fail

I didn't understand that `proptest` would test with the case where both resources have the same ID.

Then, nvm, we don't need this test as you were right, this tests the same and less cases

Reply:
> I didn't understand that `proptest` would test with the case where both resources have the same ID.

Yeah it is quite "smart" on that front and runs through a lot of different cases.

Reply:
Can you undo the commits and queue it? :)